tests/MongoDB.Driver.Core.Tests/Core/Authentication/SaslPrepHelperTests.cs

```diff
@@ -0,0 +1,153 @@
+﻿/* Copyright 2018–present MongoDB Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using FluentAssertions;
+using MongoDB.Bson.TestHelpers;
+using Xunit;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// SaslPrep unit tests.
+    /// </summary>
+    public class SaslPrepHelperTests
+    {
+        private static class SaslPrepHelperReflector
+        {
+            public static int MaxCodepoint => 
+                (int)Reflector.GetStaticFieldValue(typeof(SaslPrepHelper), nameof(MaxCodepoint));
+            
+            public static int MinCodepoint => 
+                (int)Reflector.GetStaticFieldValue(typeof(SaslPrepHelper), nameof(MinCodepoint));
+            
+            public static bool IsDefined(int codepoint) => 
+                (bool)Reflector.InvokeStatic(typeof(SaslPrepHelper), nameof(IsDefined), codepoint);
+            
+            public static bool Prohibited(int codepoint) => 
+                (bool)Reflector.InvokeStatic(typeof(SaslPrepHelper), nameof(Prohibited), codepoint);
+        }
+
+        private static readonly Lazy<int> _unAssignedCodePoint = new Lazy<int>(FindUnassignedCodePoint);
+        
+        private static int FindUnassignedCodePoint()
+        {
+            return GetLazyReversedCodePoints().First(codepoint =>
+                !SaslPrepHelperReflector.IsDefined(codepoint) && !SaslPrepHelperReflector.Prohibited(codepoint));
+        }
+
+        private static IEnumerable<int> GetLazyReversedCodePoints()
+        {
+            for (var i = SaslPrepHelperReflector.MaxCodepoint; SaslPrepHelper.MinCodepoint <= i; --i)
```

Comment:
You could have just put this for loop in FindUnassignedCodePoint and returned the first value that met the criteria.